### PR TITLE
Generate Dynamic Shapes

### PR DESCRIPTION
**Pull request by Krovatkin**
Yay!

### CODE REVIEW

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 1]
```diff
@@ -1245,9 +1274,18 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
...
               auto pttp = tensorTypeInCurrentExecutionContext(t);
-              push(stack, pttp->isSubtypeOf(expected));
+              const TypePtr& expected = af.types[inst.X];
+              auto expected_type = expected->cast<TensorType>();
+              bool pass = true;
+              if (t.defined()) {
+                pass &= bindSymbolicShapes(
+                    af, t.sizes(), expected_type->symbolic_sizes());
+                if (pass) {
+                  pttp = expected_type->merge(pttp, false);
```
Comment (zdevito) [inline torch/csrc/jit/runtime/interpreter.cpp:1285]: I think the composition of bindSymbolicShapes, merging and isSubtypeOf here is not ideal. First, it is a bit hard to follow direct intent of tracking whether the guard matches. I am still not sure what merge, and `false` is suppose to be here. Second, this is in performance critical code. `isSubtypeOf` checks are slow, and there is a lot we can do here to combine the bindSymbolicShape, and isSubtypeOf together to make a single pass over the tensor. Finally, I am confused as to whether returning false from bindSymbolicShapes will cause the guard to fail. I think it should, since we found evidence of the program no longer passing its profiled model, but it doesn't directly lead to a push `false`, so I am not sure.

**Review by zdevito — COMMENTED** [review]
Partial review: I didn't get to profiling_record.cpp/h yet, Ill try to that part tonight.

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 2]
```diff
@@ -954,6 +955,7 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
     Function** functions;
     std::function<void(std::vector<IValue>&)>* profile_functions;
     TypePtr* types;
+    ShapeSymbolTable symbols2dims;
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/runtime/interpreter.cpp:958]: ah yes... i think i actually moved it in one of the branches :-( but I guess I lost it in one of the rebases.

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 3]
```diff
@@ -21,35 +43,114 @@ ProfileOp* ProfilingRecord::createProfileNode(
...
+    auto symbol = *sym_shapes[i];
+    TORCH_INTERNAL_ASSERT(new_sizes[i]->is_static());
+    Dimension new_size = new_sizes[i]->static_size();
+    GRAPH_DEBUG("Merging symbol ", symbol);
+    if (!symbol_table.isBound(symbol)) {
+      symbol_table.assign(symbol, new_size);
+      GRAPH_DEBUG(symbol, " is now bound to ", new_size);
+      new_symbols.push_back(symbol);
+    } else {
+      if (symbol_table.getValue(symbol) == new_sizes[i]->static_size()) {
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.cpp:96]: use `new_size`

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 4]
```diff
@@ -1,12 +1,34 @@
...
 #include <torch/csrc/jit/runtime/interpreter.h>
+#include <ostream>
 
 namespace torch {
 namespace jit {
 
+c10::ShapeSymbol ShapeSymbolTable::toSymbol(
+    Dimension val,
+    std::map<Dimension, c10::ShapeSymbol>& dims2symbols,
+    ProfilingRecord* pr) {
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.cpp:13]: pr is unused here

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 5]
```diff
@@ -1,12 +1,34 @@
...
+    return new_sym;
+  }
+
+  return dims2symbols[val];
+}
+
+c10::ShapeSymbol ShapeSymbolTable::GetSymbolInSet(
+    Dimension new_size,
+    c10::ShapeSymbol set,
+    ProfilingRecord* pr) {
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.cpp:26]: pr is unused here

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 6]
```diff
@@ -1,12 +1,34 @@
...
-#include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/jit_log.h>
 #include <torch/csrc/jit/runtime/graph_executor.h>
 #include <torch/csrc/jit/runtime/interpreter.h>
+#include <ostream>
 
 namespace torch {
 namespace jit {
 
+c10::ShapeSymbol ShapeSymbolTable::toSymbol(
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.cpp:10]: I don't understand the name `toSymbol`, `getOrCreateSymbolForDim`? Shouldn't this be a method on whatever owns dims2symbols? It is weird that this is taking the map as an argument.

File: torch/csrc/jit/runtime/profiling_record.h [hunk 7]
```diff
@@ -27,6 +81,58 @@ struct ProfilingRecord {
   std::shared_ptr<Graph> profiled_graph_;
   std::mutex mutex_;
   size_t profiling_count_;
+  std::map<int64_t, std::map<Value*, TensorTypePtr>> profiled_types_per_frame_;
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.h:84]: I can't follow this data structure in the code. There are no comments about what it stores, and it results in a lot of "x.first" and "x.second" confusion.

File: torch/csrc/jit/runtime/profiling_record.h [hunk 8]
```diff
@@ -8,12 +8,66 @@
...
 
 using ::c10::TensorTypePtr;
+using Dimension = int64_t;
+
+struct ProfilingRecord;
+struct ShapeSymbolSets {
+  void reset() {
+    sets_.clear();
+  };
+  std::map<c10::ShapeSymbol, std::map<Dimension, c10::ShapeSymbol>> sets_;
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.h:25]: I don't understand what this stores, it is a nested map of ShapeSymbol to Dimension to Shape symbol?

**Review by zdevito — COMMENTED** [review]
This is getting there -- inline comments highlight a few places where I can't follow the data structures making the logic difficult to understand.

File: torch/csrc/jit/runtime/interpreter.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/interpreter.cpp:1258]: >  it doesn't directly lead to a push false, so I am not sure.

yes it's not too obvious that `expected_type` either contains static or symbolic symbols and `isSubtypeOf` will fail when comparing with `pttp` which only has static symbols unless they are either the same or we run `bindSymbolicShapes` . I explicitly added `&& bound_successfully &&` to push.

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 9]
```diff
@@ -1245,9 +1274,18 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
               push(stack, true);
             } else {
               auto t = stack.back().toTensor();
-              const TypePtr& expected = af.types[inst.X];
               auto pttp = tensorTypeInCurrentExecutionContext(t);
-              push(stack, pttp->isSubtypeOf(expected));
+              const TypePtr& expected = af.types[inst.X];
+              auto expected_type = expected->cast<TensorType>();
+              bool pass = true;
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/runtime/interpreter.cpp:1254]: agreed this can be simplified.

File: torch/csrc/jit/runtime/profiling_record.h (hunk 8 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_record.h:25]: I added a comment.
We are implementing the algorithm we discussed but the implementation merges by every dimension location, not by a whole set at once. It makes the implementation so much simpler, but it's a bit of a break from the conceptual algorithm we discussed. 
```
// A helper structure used for partitioning
// a set associated with ShapeSymbol into
// subsets each associated with a unique
// Dimension (dimension value)
// if the same symbol is
// assigned to multiple `Dimension` in one of the
// profiling runs the set of dimension locations
// associated with the set will be split into
// two subsets each associated with the new Dimension
// value 
struct ShapeSymbolSets {
```

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 10]
```diff
@@ -1,12 +1,40 @@
...
+  }
+  if (*sym_shapes.size() != new_sizes.size()) {
+    return false;
+  }
+  for (size_t i = 0; i < new_sizes.size(); i++) {
+    if (!sym_shapes[i].has_value()) {
+      continue;
+    }
+    auto symbol = *sym_shapes[i];
+    if (!isBound(symbol)) {
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.cpp:26]: This is in the hot path, so be careful about hash table lookups, this is doing two lookups even though it only needs to be doing 1.

File: torch/csrc/jit/runtime/profiling_record.h [hunk 11]
```diff
@@ -8,12 +8,127 @@
...
+    if (s.is_static()) {
+      return s.static_size();
+    }
+    return data_[s];
+  }
+  void assign(c10::ShapeSymbol s, Dimension v) {
+    TORCH_INTERNAL_ASSERT(!s.static_size());
+    data_[s] = v;
+  }
+  std::map<c10::ShapeSymbol, Dimension> data_;
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.h:117]: unordered map? A red-black tree along the fast path is going to be pretty slow.

File: torch/csrc/jit/runtime/profiling_record.h [hunk 12]
```diff
@@ -8,12 +8,127 @@
...
+      }
+      return set;
+    }
+};
+
+// ShapeSymbolTable is used by Interpreter
+// to assign dimension values to ShapeSymbols
+// and fail a guard if the same symbol
+// is assigned more than one dimension value.
+struct ShapeSymbolTable {
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.h:94]: A class used by the interpreter but not specific to profiling should probably go in a different file.

File: torch/csrc/jit/runtime/profiling_record.h [hunk 13]
```diff
@@ -8,12 +8,127 @@
...
+// * For **every** profiling run, we would like to maintain the invariant that *the same `ShapeSymbol` is always associated with the same `Dimension`*.
+// * To maintain this invariant we merge the profiling information from all profiling runs, 
+//   * For every two runs, we iterate over all `symbic_shapes_`  and compare their `ShapeSymbol`s in the same position. 
+//     * if we observe that for every dimension location that has the`ShapeSymbol S1`  in run #1 there is **only one** `ShapeSymbol S2` in the same dimension location in run #2, we conclude that the invariant holds.
+//     * However, if we observe some dimension locations in run #2 have `ShapeSymbol S2` and the other ones have `ShapeSymbol S3` we would like to partition the virtual set of dimension locations associated with `ShapeSymbol S1` into two new subsets, so the invariant holds.
+//     * The partitioning works by assigning a new symbol to the dimension locations (associated with `ShapeSymbol S1`) that have `ShapeSymbol S2` and another new symbol to the dimension locations that have `ShapeSymbol S3`. In other words,
```
Comment (zdevito) [inline torch/csrc/jit/runtime/profiling_record.h:28]: Thanks for adding this comment, it makes it much clearer what is happening in the merging algorithm.

**Review by zdevito — APPROVED** [review]
This looks good now. I have minor nits about the speed along the hot path but that can be done in a different PR since one seems functionally correct and the profiler is not on by default.
